File: Judith.NET/analysis/lexical/Token.cs
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Text;
using System.Threading.Tasks;
using static System.Runtime.InteropServices.JavaScript.JSType;
using System.Xml.Linq;

namespace Judith.NET.analysis.lexical;

public class Token {
    /// <summary>
    /// This token's type.
    /// </summary>
    [JsonConverter(typeof(StringEnumConverter))]
    public TokenKind Kind { get; init; }
    /// <summary>
    /// The string of text in the source file that was parsed as this token.
    /// </summary>
    public string Lexeme { get; init; }
    /// <summary>
    /// The position of the first character that is part of this token.
    /// </summary>
    public int Start { get; init; }
    /// <summary>
    /// The position of the first character that is not part of this token.
    /// </summary>
    public int End { get; init; }
    /// <summary>
    /// The line this token is at. If this token spans multiple lines, this
    /// number indicates the line the first character is at.
    /// </summary>
    public int Line { get; init; }
    /// <summary>
    /// The pieces of trivia that appear before this token.
    /// </summary>
    public List<Trivia> LeadingTrivia { get; init; } = [];
    /// <summary>
    /// The pieces of trivia that appear after this token.
    /// </summary>
    public List<Trivia> TrailingTrivia { get; init; } = [];

    public Token (TokenKind kind, string lexeme, int start, int end, int line) {
        Kind = kind;
        Lexeme = lexeme;
        Start = start;
        End = end;
        Line = line;
    }

    public override string ToString () {
        return $"{{{Kind}, '{Lexeme}'}}";
    }

    public static string GetTokenName (TokenKind kind) {
        return kind switch {
            TokenKind.Invalid => "Invalid",
            // Single-character tokens.
            TokenKind.Comma => "Comma",
            TokenKind.Colon => "Colon",
            TokenKind.LeftParen => "Opening parenthesis",
            TokenKind.RightParen => "Closing parenthesis",
            TokenKind.LeftCurlyBracket => "Opening curly bracket",
            TokenKind.RightCurlyBracket => "Closing curly bracket",
            TokenKind.LeftSquareBracket => "Opening square bracket",
            TokenKind.RightSquareBracket => "Closing square bracket",
            TokenKind.LeftAngleBracket => "Opening angle bracket",
            TokenKind.RightAngleBracket => "Closing angle bracket",
            TokenKind.Plus => "Plus sign",
            TokenKind.Minus => "Minus sign",
            TokenKind.Asterisk => "Asterisk",
            TokenKind.Slash => "Slash",
            TokenKind.Equal => "Equal sign",
            TokenKind.Bang => "Exclamation mark",
            TokenKind.Tilde => "Tilde",
            TokenKind.Dot => "Dot",
            TokenKind.QuestionMark => "Question mark",
            TokenKind.Pipe => "Pipe",

            // Two-character tokens.
            TokenKind.EqualEqual => "Equal equal sign",
            TokenKind.BangEqual => "Exclamation mark equal sign",
            TokenKind.TildeEqual => "Tilde equal sign",
            TokenKind.Less => "Less sign",
            TokenKind.LessEqual => "Less or equal sign",
            TokenKind.Greater => "Greater sign",
            TokenKind.GreaterEqual => "Greater or equal sign",
            TokenKind.EqualArrow => "Equal arrow",
            TokenKind.MinusArrow => "Minus arrow",
            TokenKind.DoubleColon => "Double colon",
            TokenKind.DoubleQuestionMark => "Double question mark",

            // Three-character tokens.
            TokenKind.EqualEqualEqual => "Triple equal sign",
            TokenKind.BangEqualEqual => "Not triple equal sign",

            // % operators

            // Literals
            TokenKind.Identifier => "Identifier",
            TokenKind.String => "String",
            TokenKind.Number => "Number",

            // Keywords
            TokenKind.KwConst => "const",
            TokenKind.KwVar => "var",
            TokenKind.KwTrue => "true",
            TokenKind.KwFalse => "false",
            TokenKind.KwNull => "null",
            TokenKind.KwUndefined => "undefined",
            TokenKind.KwNot => "not",
            TokenKind.KwAnd => "and",
            TokenKind.KwOr => "or",
            TokenKind.KwEnd => "end",
            TokenKind.KwIf => "if",
            TokenKind.KwElse => "else",
            TokenKind.KwElsif => "elsif",
            TokenKind.KwThen => "then",
            TokenKind.KwMatch => "match",
            TokenKind.KwLoop => "loop",
            TokenKind.KwWhile => "while",
            TokenKind.KwFor => "for",
            TokenKind.KwIn => "in",
            TokenKind.KwDo => "do",
            TokenKind.KwReturn => "return",
            TokenKind.KwYield => "yield",
            TokenKind.KwBreak => "break",
            TokenKind.KwContinue => "continue",
            TokenKind.KwGoto => "goto",
            TokenKind.KwFunc => "func",
            TokenKind.KwGenerator => "generator",
            TokenKind.KwHid => "hid",
            TokenKind.KwOper => "oper",
            TokenKind.KwTypedef => "typedef",
            TokenKind.KwStruct => "struct",
            TokenKind.KwInterface => "interface",
            TokenKind.KwClass => "class",
            TokenKind.KwPub => "pub",
            TokenKind.KwMut => "mut",
            TokenKind.KwStatic => "static",

            // Private keywords
            TokenKind.PkwPrint => "__p_print",

            // Directives

            // Other
            TokenKind.Comment => "Comment",

            TokenKind.EOF => "End of file",
            TokenKind.Error => "Error",
            _ => "Invalid token code"
        };
    }
}

public class StringToken : Token {
    public StringLiteralKind StringKind { get; init; }
    /// <summary>
    /// The character used to delimit the string.
    /// </summary>
    public char Delimiter { get; init; }
    /// <summary>
    /// The amount of characaters used to delimit the string.
    /// </summary>
    public int DelimiterCount { get; init; }
    /// <summary>
    /// The index of the first character of this string in the column it
    /// belongs to.
    /// </summary>
    public int ColumnIndex { get; init; }

    public StringToken (
        string lexeme,
        int start,
        int end,
        int line,
        StringLiteralKind stringKind,
        char delimiter,
        int delimiterCount,
        int columnIndex
    )
        : base(TokenKind.String, lexeme, start, end, line) {
        StringKind = stringKind;
        Delimiter = delimiter;
        DelimiterCount = delimiterCount;
        ColumnIndex = columnIndex;
    }

    public override string ToString () {
        return $"{{{Kind}, '{Lexeme}', Kind: {StringKind}, Delimiter: " +
            $"'{Delimiter}' (x{DelimiterCount})}}";
    }
}

File: judc/src/judith/lexical/token.rs
use std::fmt::{Debug, Formatter};
use serde::Serialize;
use crate::SourceSpan;

#[derive(Clone, Serialize)]
pub struct RegularToken {
    pub kind: TokenKind,
    pub lexeme: String,
    pub start: i64,
    pub end: i64,
    pub line: i64,
    pub leading_trivia: Vec<Trivia>,
    pub trailing_trivia: Vec<Trivia>,
}

#[derive(Clone, Serialize)]
pub struct StringToken {
    pub base: RegularToken,
    pub string_kind: StringLiteralKind,
    pub delimiter: char,
    pub delimiter_count: i32,
    pub column: i32,
}

#[derive(Debug, Clone, Serialize)]
pub enum StringLiteralKind {
    Regular,
    Raw,
}

#[derive(Debug, Clone, Serialize)]
pub enum Token {
    Regular(RegularToken),
    String(StringToken),
}

impl Token {
    pub fn kind(&self) -> TokenKind {
        match self {
            Token::Regular(t) => t.kind.clone(),
            Token::String(t) => t.base.kind.clone(),
        }
    }

    pub fn base(&self) -> &RegularToken {
        match self {
            Token::Regular(t) => t,
            Token::String(t) => &t.base,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct Trivia {
    pub kind: TriviaKind,
    pub lexeme: String,
    pub span: SourceSpan,
}

#[derive(Debug, Copy, Clone, PartialEq, Serialize)]
pub enum TokenKind {
    // Error tokens
    Invalid = 0,

    // Single-character tokens
    Comma,
    Colon,
    LeftParen,
    RightParen,
    LeftCurlyBracket,
    RightCurlyBracket,
    LeftSquareBracket,
    RightSquareBracket,
    LeftAngleBracket,
    RightAngleBracket,
    Plus,
    Minus,
    Asterisk,
    Slash,
    Equal,
    Bang,
    Tilde,
    Dot,
    QuestionMark,
    Ampersand,
    Pipe,

    // Two-character tokens.
    EqualEqual,
    BangEqual,
    TildeTilde,
    BangTilde,
    Less,
    LessEqual,
    Greater,
    GreaterEqual,
    EqualArrow,
    MinusArrow,
    DoubleColon,
    DoubleQuestionMark,

    // Three-character tokens.
    EqualEqualEqual,
    BangEqualEqual,

    // % operators

    // Identifier
    Identifier,

    // Literals
    String,
    Number,

    // Keywords
    KwTrue,
    KwFalse,
    KwNull,
    KwUndefined,
    KwNot,
    KwAnd,
    KwOr,
    KwEnd,
    KwIf,
    KwElse,
    KwElsif,
    KwMatch,
    KwThen,
    KwLoop,
    KwWhile,
    KwFor,
    KwIn,
    KwDo,
    KwReturn,
    KwYield,
    KwBreak,
    KwContinue,
    KwGoto,
    KwFunc,
    KwGenerator,
    KwOper,
    KwTypedef,
    KwStruct,
    KwInterface,
    KwClass,
    KwHid,
    KwPub,
    KwLet,
    KwFinal,
    KwMut,
    KwSh,
    KwRef,

    // Private keywords
    PkwPrint,

    // Directives

    // Other
    Comment,

    EOF,
}

impl TokenKind {
    pub fn name (&self) -> &str {
        match self {
            TokenKind::Invalid => "<invalid token>",
            TokenKind::Comma => "','",
            TokenKind::Colon => "'.'",
            TokenKind::LeftParen => "'('",
            TokenKind::RightParen => "')'",
            TokenKind::LeftCurlyBracket => "'{",
            TokenKind::RightCurlyBracket => "'}'",
            TokenKind::LeftSquareBracket => "'['",
            TokenKind::RightSquareBracket => "']'",
            TokenKind::LeftAngleBracket => "'<'",
            TokenKind::RightAngleBracket => "'>'",
            TokenKind::Plus => "'+'",
            TokenKind::Minus => "'-'",
            TokenKind::Asterisk => "'*'",
            TokenKind::Slash => "'/'",
            TokenKind::Equal => "'='",
            TokenKind::Bang => "'!'",
            TokenKind::Tilde => "'~'",
            TokenKind::Dot => "'.'",
            TokenKind::QuestionMark => "'?'",
            TokenKind::Ampersand => "'&'",
            TokenKind::Pipe => "'|'",
            TokenKind::EqualEqual => "'=='",
            TokenKind::BangEqual => "'!='",
            TokenKind::TildeTilde => "'~~'",
            TokenKind::BangTilde => "'!~'",
            TokenKind::Less => "'<'",
            TokenKind::LessEqual => "'<='",
            TokenKind::Greater => "'>'",
            TokenKind::GreaterEqual => "'>='",
            TokenKind::EqualArrow => "'=>'",
            TokenKind::MinusArrow => "'->'",
            TokenKind::DoubleColon => "'::'",
            TokenKind::DoubleQuestionMark => "'??'",
            TokenKind::EqualEqualEqual => "'==='",
            TokenKind::BangEqualEqual => "'!=='",
            TokenKind::Identifier => "<identifier>",
            TokenKind::String => "<string literal>",
            TokenKind::Number => "<number literal>",
            TokenKind::KwTrue => "'true'",
            TokenKind::KwFalse => "'false'",
            TokenKind::KwNull => "'null'",
            TokenKind::KwUndefined => "'undefined'",
            TokenKind::KwNot => "'not'",
            TokenKind::KwAnd => "'and'",
            TokenKind::KwOr => "'or'",
            TokenKind::KwEnd => "'end'",
            TokenKind::KwIf => "'if'",
            TokenKind::KwElse => "'else'",
            TokenKind::KwElsif => "'elsif'",
            TokenKind::KwMatch => "'match'",
            TokenKind::KwThen => "'then'",
            TokenKind::KwLoop => "'loop'",
            TokenKind::KwWhile => "'while'",
            TokenKind::KwFor => "'for'",
            TokenKind::KwIn => "'in'",
            TokenKind::KwDo => "'do'",
            TokenKind::KwReturn => "'return'",
            TokenKind::KwYield => "'yield'",
            TokenKind::KwBreak => "'break'",
            TokenKind::KwContinue => "'continue'",
            TokenKind::KwGoto => "'goto'",
            TokenKind::KwFunc => "'func'",
            TokenKind::KwGenerator => "'generator'",
            TokenKind::KwOper => "'oper'",
            TokenKind::KwTypedef => "'typedef'",
            TokenKind::KwStruct => "'struct'",
            TokenKind::KwInterface => "'interface'",
            TokenKind::KwClass => "'class'",
            TokenKind::KwHid => "'hid'",
            TokenKind::KwPub => "'pub'",
            TokenKind::KwLet => "'let'",
            TokenKind::KwFinal => "'final'",
            TokenKind::KwMut => "'mut'",
            TokenKind::KwSh => "'sh'",
            TokenKind::KwRef => "'ref'",
            TokenKind::PkwPrint => "'__p_print'",
            TokenKind::Comment => "<comment>",
            TokenKind::EOF => "<end of file>",
        }
    }
}

#[derive(Debug, Clone, PartialEq, Serialize)]
pub enum TriviaKind {
    SingleLineComment,
    MultiLineComment,
    Whitespace,
    LineBreak,
    Directive,
}

impl Debug for RegularToken {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("RegularToken")
            .field("kind", &self.kind)
            .field("lexeme", &self.lexeme)
            .field("start", &self.start)
            .field("end", &self.end)
            .field("line", &self.line)
            .field("leading_trivia", &self.leading_trivia.len())
            .field("trailing_trivia", &self.trailing_trivia.len())
            .finish()
    }
}

impl Debug for StringToken {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("StringToken")
            .field("base", &self.base)
            .field("string_kind", &self.string_kind)
            .field("delimiter", &self.delimiter)
            .field("delimiter_count", &self.delimiter_count)
            .field("column", &self.column)
            .finish()
    }
}